File: python/codewire/http.py
from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

from codewire.errors import CodewireError


class HttpClient:
    def __init__(self, base_url: str, api_key: str, org_id: str):
        self._client = httpx.Client(
            base_url=base_url,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
        )
        self.org_id = org_id

    def resolve_org_id(self) -> str:
        if not self.org_id:
            raise ValueError(
                "org_id is required — pass org_id to Codewire() "
                "or set CODEWIRE_ORG_ID"
            )
        return self.org_id

    def org_path(self, pattern: str) -> str:
        return f"/organizations/{self.resolve_org_id()}{pattern}"

    def request(self, method: str, path: str, body: Any = None) -> Any:
        resp = self._client.request(method, path, json=body)
        if resp.status_code >= 400:
            detail = resp.reason_phrase or ""
            try:
                data = resp.json()
                detail = data.get("detail") or data.get("message") or detail
            except Exception:
                pass
            raise CodewireError(resp.status_code, detail)
        if resp.status_code == 204:
            return None
        return resp.json()

    def request_raw(
        self,
        method: str,
        path: str,
        data: bytes | None = None,
        content_type: str | None = None,
    ) -> httpx.Response:
        headers: dict[str, str] = {}
        if content_type:
            headers["Content-Type"] = content_type
        resp = self._client.request(method, path, content=data, headers=headers)
        if resp.status_code >= 400:
            detail = resp.reason_phrase or ""
            try:
                d = resp.json()
                detail = d.get("detail") or d.get("message") or detail
            except Exception:
                pass
            raise CodewireError(resp.status_code, detail)
        return resp

    def get(self, path: str) -> Any:
        return self.request("GET", path)

    def post(self, path: str, body: Any = None) -> Any:
        return self.request("POST", path, body)

    def put(self, path: str, body: Any = None) -> Any:
        return self.request("PUT", path, body)

    def patch(self, path: str, body: Any = None) -> Any:
        return self.request("PATCH", path, body)

    def delete(self, path: str) -> Any:
        return self.request("DELETE", path)

    def close(self) -> None:
        self._client.close()

File: typescript/src/http.ts
import { CodewireError } from "./types.js";

export class HttpClient {
  readonly orgId: string;

  constructor(
    private baseUrl: string,
    private apiKey: string,
    orgId: string,
  ) {
    this.orgId = orgId;
  }

  resolveOrgId(): string {
    if (!this.orgId) {
      throw new Error(
        "orgId is required — pass orgId in ClientOptions or set CODEWIRE_ORG_ID",
      );
    }
    return this.orgId;
  }

  orgPath(pattern: string): string {
    return `/organizations/${this.resolveOrgId()}${pattern}`;
  }

  private headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
    };
  }

  async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const res = await fetch(url, {
      method,
      headers: this.headers(),
      body: body ? JSON.stringify(body) : undefined,
    });

    if (!res.ok) {
      let detail = res.statusText;
      try {
        const err = await res.json();
        detail = err.detail ?? err.message ?? detail;
      } catch {
        // ignore parse errors
      }
      throw new CodewireError(res.status, detail);
    }

    if (res.status === 204) return undefined as T;
    return res.json() as Promise<T>;
  }

  async requestRaw(
    method: string,
    path: string,
    body?: unknown,
    contentType?: string,
  ): Promise<Response> {
    const url = `${this.baseUrl}${path}`;
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
    };
    if (contentType) headers["Content-Type"] = contentType;

    const res = await fetch(url, {
      method,
      headers,
      body: body as BodyInit | undefined,
    });
    if (!res.ok) {
      let detail = res.statusText;
      try {
        const err = await res.json();
        detail = err.detail ?? err.message ?? detail;
      } catch {
        // ignore
      }
      throw new CodewireError(res.status, detail);
    }
    return res;
  }

  get<T>(path: string): Promise<T> {
    return this.request("GET", path);
  }

  post<T>(path: string, body?: unknown): Promise<T> {
    return this.request("POST", path, body);
  }

  put<T>(path: string, body?: unknown): Promise<T> {
    return this.request("PUT", path, body);
  }

  patch<T>(path: string, body?: unknown): Promise<T> {
    return this.request("PATCH", path, body);
  }

  delete<T>(path: string): Promise<T> {
    return this.request("DELETE", path);
  }
}
